src/analyses/base.ml

```diff
@@ -2168,11 +2167,31 @@ struct
     Printable.get_short_list (GU.demangle f.svar.vname ^ "(") ")" args_short
 
   let threadenter ctx (lval: lval option) (f: varinfo) (args: exp list): D.t list =
-    try
-      [make_entry ~thread:true ctx f args]
-    with Not_found ->
+    match Cilfacade.getdec f with
+    | fd ->
+      [make_entry ~thread:true ctx fd args]
+    | exception Not_found ->
       (* Unknown functions *)
-      [ctx.local]
+      (* Copied from special unknown invalidate *)
+      let st = ctx.local in
+      let addrs =
+        if get_bool "sem.unknown_function.invalidate.globals" then (
+          M.warn_each "INVALIDATING ALL GLOBALS!";
+          foldGlobals !Cilfacade.current_file (fun acc global ->
+              match global with
+              | GVar (vi, _, _) when not (is_static vi) ->
+                mkAddrOf (Var vi, NoOffset) :: acc
+              (* TODO: what about GVarDecl? *)
+              | _ -> acc
+            ) []
+        )
+        else
+          []
+      in
+      (* TODO: what about escaped local variables? *)
+      (* invalidate non-static globals for unknown functions *)
+      let st = invalidate ~ctx (Analyses.ask_of_ctx ctx) ctx.global st addrs in
+      [st]
```

Comment:
Can one deduplicate this code?

Reply:
Should be doable, but the bigger question is actually how much of it should be shared. `special` also has special more precise invalidation for `LibraryFunctions` and a bunch of special implementations. One option might even be just delegating to `special` here, but I don't know if that has any value. Spawning a thread from `memcpy` or `malloc` seems like complete nonsense.

Reply:
I extracted just this part out. Calling `special` from `threadenter` is problematic because it would want to recursively spawn threads, but that's not currently allowed in `threadenter`.